presto-cli/src/main/java/io/prestosql/cli/OutputHandler.java

```diff
@@ -13,85 +13,95 @@
  */
 package io.prestosql.cli;
 
-import io.airlift.units.Duration;
+import com.google.common.collect.Queues;
 import io.prestosql.client.StatementClient;
 
 import java.io.Closeable;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import static io.airlift.units.Duration.nanosSince;
 import static java.util.Collections.unmodifiableList;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
 public final class OutputHandler
         implements Closeable
 {
-    private static final Duration MAX_BUFFER_TIME = new Duration(3, SECONDS);
-    private static final int MAX_BUFFERED_ROWS = 10_000;
+    private static final int MAX_QUEUED_ROWS = 50_000;
+    private static final int MIN_BUFFERED_ROWS = 100;
+    private static final long MAX_BUFFER_TIMEOUT = 3;
 
     private final AtomicBoolean closed = new AtomicBoolean();
-    private final List<List<?>> rowBuffer = new ArrayList<>(MAX_BUFFERED_ROWS);
+    private final BlockingQueue<List<?>> rowQueue = new LinkedBlockingQueue<>(MAX_QUEUED_ROWS);
     private final OutputPrinter printer;
+    private List<List<?>> rowBuffer = new ArrayList<>();
 
-    private long bufferStart;
+    private CompletableFuture<Void> future;
 
     public OutputHandler(OutputPrinter printer)
     {
         this.printer = requireNonNull(printer, "printer is null");
     }
 
-    public void processRow(List<?> row)
-            throws IOException
-    {
-        if (rowBuffer.isEmpty()) {
-            bufferStart = System.nanoTime();
-        }
-
-        rowBuffer.add(row);
-        if (rowBuffer.size() >= MAX_BUFFERED_ROWS) {
-            flush(false);
-        }
-    }
-
     @Override
     public void close()
             throws IOException
     {
         if (!closed.getAndSet(true)) {
-            flush(true);
             printer.finish();
         }
     }
 
     public void processRows(StatementClient client)
             throws IOException
     {
-        while (client.isRunning()) {
-            Iterable<List<Object>> data = client.currentData().getData();
-            if (data != null) {
-                for (List<Object> row : data) {
-                    processRow(unmodifiableList(row));
+        this.future = CompletableFuture.runAsync(() -> {
+            while (client.isRunning()) {
+                Iterable<List<Object>> data = client.currentData().getData();
+                if (data != null) {
+                    for (List<Object> row : data) {
```

Comment:
Ideally, we should put the entire batch in the queue, rather than extracting the individual rows, but this is fine for now.